src/software/amazon/ion/impl/bin/AbstractIonWriter.java

```diff
@@ -113,8 +113,23 @@ public final void writeValueRecursive(final IonReader reader) throws IOException
                 writeBool(booleanValue);
                 break;
             case INT:
-                final BigInteger bigIntegerValue = reader.bigIntegerValue();
-                writeInt(bigIntegerValue);
+                switch (reader.getIntegerType())
```

Comment:
Tried my hand at a callback. Not sure if it actually saves the user any effort or not, since they still need to implement a "visit" method for each integer type. But it does get rid of `IntegerType` altogether. Let me know if it's along the lines of what you were thinking.